File: git/scripts/pre-push.py
#!/usr/bin/env python3

import os
import subprocess
import sys

def log(message, level="INFO"):
    # ANSI escape sequences for colored output
    levels = {
        "DEBUG": "\033[34m",    # Blue
        "INFO": "\033[32m",     # Green
        "WARNING": "\033[33m",  # Yellow
        "ERROR": "\033[31m",    # Red
        "CRITICAL": "\033[35m", # Magenta
        "RESET": "\033[0m"      # Reset
    }
    color_code = levels.get(level, levels["RESET"])
    print(f"{color_code}[{level}] {message}{levels['RESET']}")

def is_executable(file_path):
    # On Windows, check if the file exists
    # On Unix-like systems, check if the file is executable
    return os.path.isfile(file_path) and (os.name == 'nt' or os.access(file_path, os.X_OK))

def read_single_line(file_path):
    with open(file_path, 'r') as file:
        line = file.readline().strip()
    return line

def get_parent_directory(path):
    return os.path.abspath(os.path.join(path, os.pardir))

def main():
    worktree_dir = subprocess.check_output(['git', 'rev-parse', '--show-toplevel']).strip().decode()
    branches_dir = get_parent_directory(worktree_dir)
    repo_dir = get_parent_directory(branches_dir)
    hooks_dir = os.path.join(repo_dir, 'hooks')

    # This works because the list of refs being pushed is passed on standard input
    if 'null' in sys.stdin.read():
        log("You really do not want to push this branch. Aborting.", "ERROR")
        sys.exit(1)

    # Check if the hook exists in the worktree hooks directory
    pre_push_hook = os.path.join(hooks_dir, 'pre-push')
    if is_executable(pre_push_hook):
        log(f"Found {pre_push_hook}. Executing.")
        # Execute the worktree hook
        os.execv(pre_push_hook, [pre_push_hook] + sys.argv[1:])
    else:
        # Execute the global hook
        log(f"Completed pre-push. No hook under {pre_push_hook} found. Exiting.")

if __name__ == "__main__":
    main()

File: git/scripts/pre-push.hs
#!/usr/bin/env stack
-- stack script --resolver lts-23.8
-- package directory
-- package process
-- package text
-- package base
-- package containers
{-# LANGUAGE OverloadedStrings #-}

-- import System.Directory (doesFileExist, getPermissions, executable)
-- import System.Environment (getArgs)
-- import System.Exit (exitFailure)
-- import System.FilePath (takeDirectory, (</>))
-- import System.IO (hGetContents, stdin)
-- import System.Process (readProcess, callProcess, callCommand)
-- import Control.Monad (when, unless)
-- import Data.List (isInfixOf)
-- import Data.Text (Text, pack, unpack)
-- import qualified Data.Text as T
-- import qualified Data.Text.IO as TIO
-- import Text.Printf (printf)
-- import Data.Monoid ((<>))
import System.Directory (doesFileExist, getPermissions, executable)
import System.Environment (getArgs)
import System.Exit (exitFailure)
import System.FilePath (takeDirectory, (</>))
import System.IO (hGetContents, stdin)
import System.Process (readProcess, callProcess, callCommand)
import Control.Monad (when, unless)
import Data.List (isInfixOf)
import Data.Text (Text, pack, unpack)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Text.Printf (printf)
import Data.Monoid ((<>))

-- logMessage :: Text -> Text -> IO ()
-- logMessage level message = do
--     let levels = [("DEBUG", "\ESC[34m"), ("INFO", "\ESC[32m"), ("WARNING", "\ESC[33m"), ("ERROR", "\ESC[31m"), ("CRITICAL", "\ESC[35m"), ("RESET", "\ESC[0m")]
--     let colorCode = maybe "\ESC[0m" id (lookup level levels)
--     TIO.putStrLn $ pack $ printf "%s[%s] %s%s" colorCode (unpack level) (unpack message) (snd $ last levels)
logMessage :: Text -> Text -> IO ()
logMessage level message = do
    let levels :: [(Text, Text)]
        levels = [("DEBUG", "\ESC[34m"), ("INFO", "\ESC[32m"), ("WARNING", "\ESC[33m"), ("ERROR", "\ESC[31m"), ("CRITICAL", "\ESC[35m"), ("RESET", "\ESC[0m")]
    let colorCode = maybe "\ESC[0m" id (lookup level levels)
    TIO.putStrLn $ colorCode <> "[" <> level <> "] " <> message <> (snd $ last levels)

isExecutable :: FilePath -> IO Bool
isExecutable filePath = do
    exists <- doesFileExist filePath
    if exists
        then do
            perms <- getPermissions filePath
            return $ executable perms
        else return False

readSingleLine :: FilePath -> IO Text
readSingleLine filePath = do
    contents <- TIO.readFile filePath
    return $ head $ T.lines contents

getParentDirectory :: FilePath -> FilePath
getParentDirectory = takeDirectory . takeDirectory

main :: IO ()
main = do
    -- let binaryName = "pre-push-binary"
    -- binaryExists <- doesFileExist binaryName
    -- unless binaryExists $ do
    --     logMessage "INFO" "Binary does not exist. Compiling..."
    --     callCommand "stack ghc -- -o pre-push-binary script.hs"
    --     logMessage "INFO" "Compilation finished."

    worktreeDir <- fmap init $ readProcess "git" ["rev-parse", "--show-toplevel"] ""
    let branchesDir = getParentDirectory worktreeDir
    let repoDir = getParentDirectory branchesDir
    let hooksDir = repoDir </> "hooks"

    stdinContent <- hGetContents stdin
    when ("null" `isInfixOf` stdinContent) $ do
        logMessage "ERROR" "You really do not want to push this branch. Aborting."
        exitFailure

    let prePushHook = hooksDir </> "pre-push"
    hookExists <- isExecutable prePushHook
    if hookExists
        then do
            logMessage "INFO" $ "Found " <> pack prePushHook <> ". Executing."
            args <- getArgs
            callProcess prePushHook args
        else
            logMessage "INFO" $ "Completed pre-push. No hook under " <> pack prePushHook <> " found. Exiting."
